### ISSUE TITLE
Handling of macro's in abap code (including classes)

### ISSUE THREAD

**Issue by ed-holland** [abaplint/abaplint#1884]
Background (to my knowledge):

The macro definitions are inserted "Just in Time" inside the abap code being executed.
They can be local (to the form, function, class) or global definitions or a combination.
The Macro has replacements (&1...&x). These are replaced before the macro is inserted into the macro call location.
The expanded code is then parsed and executed.
Note that Macro calls can be nested!

Example:
Global definition:
![image](https://user-images.githubusercontent.com/53995991/113295822-fa129a80-92f8-11eb-8f31-755741fe6590.png)
Local definition:
![image](https://user-images.githubusercontent.com/53995991/113295901-13b3e200-92f9-11eb-8594-74d86cbbae4a.png)

If linter tries to parse without expanding as described above, rules are parsed wrongly.
Example is indenting (especially if macro contains incomplete loop/case statements).
```
./src/ecc/#jcs#cl_fi_badi_acc_document.clas.abap[27, 3]  - Indentation problem, expected 4 spaces (indentation) [E]
```

Macro coding is not a technique to be used regularly or lightly. In this case, it allows me to provide the same coding for a BADI / BTE integration although the API is quite different. I might rewrite it if there is time in the future.

**Comment by larshp** [comment]
ugh, thats a nightmare

the below will give a false positive `Variable "foo" not used`

```abap
DEFINE sub.
write foo.
END-OF-DEFINITION.
DEFINE top.
sub.
END-OF-DEFINITION.
FORM moo.
  DATA foo TYPE c LENGTH 1.
  top.
ENDFORM.
```

**Comment by ed-holland** [comment]
Yep, you can do interesting stuff though when using variables ;-)

**Comment by ed-holland** [comment]
Example:
![image](https://user-images.githubusercontent.com/53995991/114303219-5631a780-9acd-11eb-84b6-f402cfa4b133.png)

**Comment by larshp** [comment]
yea, quite similar to the testcases added https://github.com/abaplint/abaplint/pull/1906/files#diff-81fcb78403199180e669b8c6b7391dbfed0f6eb0cd465ff5f040de8643e2325eR56

PR https://github.com/abaplint/abaplint/pull/1906 should help with expanding macros

**Comment by ed-holland** [comment]
Maybe add another testcase with linked language elements like CASE/END-CASE, LOOP/END-LOOP, IF/ENDIF etc.

**Comment by larshp** [comment]
its already published, try in ~5 minutes on the https://playground.abaplint.org

**Comment by ed-holland** [comment]
The result is very good from what i can see in my test. The correct line is identified.
```
REPORT zfoobar.

DATA: foo_1(1) TYPE c.

DEFINE sub.
  WRITE foo_&1.
END-OF-DEFINITION.
DEFINE top.
  sub &1.
END-OF-DEFINITION.
FORM moo.
  DATA foo_0 TYPE c LENGTH 1.
  top 0.
  top 1.
  top 2.
ENDFORM.

DEFINE do_loop.
  LOOP AT &1 ASSIGNING FIELD-SYMBOL(&2).
END-OF-DEFINITION.

DEFINE do_end.
  WRITE 'bar'.
  END&1.
END-OF-DEFINITION.

START-OF-SELECTION.

  FIELD-SYMBOLS: <ls>.
  TABLES: t000.
  DATA ltb LIKE t000 OCCURS 0 WITH HEADER LINE.
  SELECT * FROM t100 INTO TABLE ltb.

FORM foo.
  do_loop ltb <ls>.
  WRITE: / 'Test'.
  do_end loop.
ENDFORM.
```
The errors are the same as what SAP provided. (FOO_2 does not exist).
Even the tricky endloop statement was handled.

**Comment by larshp** [comment]
😁

**Comment by ed-holland** [comment]
Just did a bigger test with my major projects (reduced global macro/constants in profile).
Most of the macro's and most of the constants could be removed.
The macros that stayed are related to their code not being 700 compliant (checking from 740 dependencies).

Very happy with the results :-)
